homeassistant/components/nanoleaf/light.py

```diff
@@ -227,8 +227,10 @@ def update(self):
             self._available = self._light.available
             self._brightness = self._light.brightness
             self._color_temp = self._light.color_temperature
-            self._effect = self._light.effect
             self._effects_list = self._light.effects
+            self._effect = (
+                self._light.effect if self._light.effect in self._effects_list else None
```

Comment:
The correct fix is to check that the effect that come in the service call data to the turn_on method is in `self._effects_list` and should be done here:
https://github.com/home-assistant/core/blob/de7bbd3e24bee53b98a5481ac6fecf9505ca0565/homeassistant/components/nanoleaf/light.py#L212

Log an error and return or raise an error.

Reply:
That's where I originally tried the check, but it doesn't appear to work. 

Stepping through the code, it looks like `effect` represents what the user manually selects from the Effects drop-down on the color-wheel UI, and `self._light.effect` represents what the Nanoleaf API returns for the light's current state.

It's Nanoleaf's official API that's untrustworthy here, the drop-down never contains the invalid "\*Solid\*" option, and doesn't appear to need any further checks.

Reply:
This is not about the drop down. The scene will call a service which calls the corresponding entity method `turn_on`.

Reply:
As for whether to fix this in the pynanoleaf library that consumes the nanoleaf api, I obviously defer to the philosophies of experienced HA devs like yourself, but I generally think of wrapper libraries as "pure" deliverers of the API they consume, and that clients should be responsible for tweaking the data when the format doesn't match the client's preferences.

That said, I realize HA has its own philosophies and conventions, so I'm happy to submit a PR to pynanoleaf if you think that's the proper place!

Reply:
Ahh, got it. I'll give that a shot then.

Reply:
I think we can fix it here for now.

Reply:
Well, your suggested change does fix the issue, but it breaks again when I `raise ValueError`. Essentially just trading one error for another.

Does that match your expectation? Would it suffice for us to log a warning/error rather than raise the exception? 

(Note that with your change, the generated scenes.yaml still contains the invalid `effect: '*Solid*'`, it's just that we're blocking it from being sent to the Nanoleaf API)

Here's my code:
```
if effect:
  if effect in self._effects_list:
    self._light.effect = effect
  else:
    raise ValueError(f"Attempting to apply non-existent effect '{effect}'")
```

Reply:
Raising the `ValueError` in the entity actuator method should show as a message in the frontend and a log message stack trace.

I think that's good.

To avoid the invalid scene I think we should also keep the fix on this line here.

Please add a code comment why we need this check.

Reply:
If the library is fixed we can remove the check with the comment.

Reply:
Like this? (See pushed changes)